modules/host/network.go

```diff
@@ -174,6 +161,20 @@ func (h *Host) threadedListen() {
 	}
 	defer h.tg.Done()
 
+	// Automatically close the listener when h.tg.Stop() is called.
+	err = h.tg.Add()
+	if err != nil {
+		return
+	}
+	go func() {
+		defer h.tg.Done()
+		<-h.tg.StopChan()
+		err = h.listener.Close()
+		if err != nil {
+			h.log.Println("Closing the listener failed:", err)
+		}
+	}()
+
```

Comment:
I'm sorry I'm still failing to see how this reintroduces the bug.

You bring up a good point about placing the shutdown of the listener next to the creation of the listener, rather than the `listener.Accept`.

This pattern is in contrast to the shutdown of the `conn` created by `listener.Accept`. If we were to follow this pattern we would move `conn.Close` to `threadedListen` like so
```go
// ...
conn, err := h.listener.Accept()
if err != nil {
    return
}

// Close the conn on host.Close or when threadedHandleConn terminates,
// whichever comes first.
connCloseChan := make(chan struct{})
go func() {
    select {
    case <-h.tg.StopChan():
    case <-connCloseChan:
    }
    conn.Close()
}()

go func() {
    h.threadedHandleConn(conn)
    close(connCloseChan)
}()
// ...
```
It's different in that it is possible and perfectly fine for multiple goroutines to call and wait on `listener.Accept` simultaneously, but it is probably an error if multiple goroutines are calling `conn.Read` or `conn.Write` simultaneously.

NACK agreed.